File: src/main/java/com/thebyteslayer/tracker/CompassListener.java
package com.thebyteslayer.tracker;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.player.PlayerItemHeldEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

public class CompassListener implements Listener {
    
    private final CompassTracker compassTracker;
    
    public CompassListener(CompassTracker compassTracker) {
        this.compassTracker = compassTracker;
    }
    
    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        Player player = event.getPlayer();
        checkPlayerInventory(player);
    }
    
    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        if (!(event.getWhoClicked() instanceof Player)) {
            return;
        }
        
        ItemStack item = event.getCurrentItem();
        if (item != null && item.getType() == Material.COMPASS) {
            Player player = (Player) event.getWhoClicked();
            scheduleCompassCheck(player);
        }
    }
    
    @EventHandler
    public void onPlayerItemHeld(PlayerItemHeldEvent event) {
        Player player = event.getPlayer();
        ItemStack item = player.getInventory().getItem(event.getNewSlot());
        
        if (item != null && item.getType() == Material.COMPASS) {
            scheduleCompassCheck(player);
        }
    }
    
    private void scheduleCompassCheck(Player player) {
        compassTracker.plugin.getServer().getScheduler().runTaskLater(
            compassTracker.plugin, 
            () -> checkPlayerInventory(player), 
            1L
        );
    }
    
    private void checkPlayerInventory(Player player) {
        for (ItemStack item : player.getInventory().getContents()) {
            if (item != null && item.getType() == Material.COMPASS) {
                checkCompass(player, item);
            }
        }
        
        ItemStack mainHand = player.getInventory().getItemInMainHand();
        if (mainHand.getType() == Material.COMPASS) {
            checkCompass(player, mainHand);
        }
        
        ItemStack offHand = player.getInventory().getItemInOffHand();
        if (offHand.getType() == Material.COMPASS) {
            checkCompass(player, offHand);
        }
    }
    
    private void checkCompass(Player holder, ItemStack compass) {
        if (!compass.hasItemMeta() || !compass.getItemMeta().hasDisplayName()) {
            return;
        }
        
        String displayName = compass.getItemMeta().getDisplayName();
        String targetName = ChatColor.stripColor(displayName);
        
        Player target = compassTracker.plugin.getServer().getPlayer(targetName);
        if (target != null && target.isOnline()) {
            compassTracker.addCompassTarget(holder.getUniqueId(), targetName);
            
            ItemMeta meta = compass.getItemMeta();
            if (meta.getLore() == null || !meta.getLore().contains(ChatColor.GRAY + "Tracking: " + targetName)) {
                meta.setLore(java.util.Arrays.asList(
                    ChatColor.GRAY + "Tracking: " + targetName,
                    ChatColor.DARK_GRAY + "Tracker Compass"
                ));
                compass.setItemMeta(meta);
            }
        }
    }
} 

File: src/main/kotlin/com/thebyteslayer/mc/tracker/CompassListener.kt
package com.thebyteslayer.mc.tracker

import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.player.PlayerItemHeldEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.inventory.ItemStack

class CompassListener(private val compassTracker: CompassTracker) : Listener {

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        val player = event.player
        checkPlayerInventory(player)
    }

    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        val player = event.whoClicked as? Player ?: return

        val item = event.currentItem
        if (item != null && item.type == Material.COMPASS) {
            scheduleCompassCheck(player)
        }
    }

    @EventHandler
    fun onPlayerItemHeld(event: PlayerItemHeldEvent) {
        val player = event.player
        val item = player.inventory.getItem(event.newSlot)

        if (item != null && item.type == Material.COMPASS) {
            scheduleCompassCheck(player)
        }
    }

    private fun scheduleCompassCheck(player: Player) {
        compassTracker.plugin.server.scheduler.runTaskLater(
            compassTracker.plugin,
            Runnable { checkPlayerInventory(player) },
            1L
        )
    }

    private fun checkPlayerInventory(player: Player) {
        for (item in player.inventory.contents) {
            if (item != null && item.type == Material.COMPASS) {
                checkCompass(player, item)
            }
        }

        val mainHand = player.inventory.itemInMainHand
        if (mainHand.type == Material.COMPASS) {
            checkCompass(player, mainHand)
        }

        val offHand = player.inventory.itemInOffHand
        if (offHand.type == Material.COMPASS) {
            checkCompass(player, offHand)
        }
    }

    private fun checkCompass(holder: Player, compass: ItemStack) {
        val meta = compass.itemMeta ?: return
        if (!meta.hasDisplayName()) {
            return
        }

        val displayName = meta.displayName ?: return
        val targetName = ChatColor.stripColor(displayName)!!

        val target = compassTracker.plugin.server.getPlayer(targetName)
        if (target != null && target.isOnline) {
            compassTracker.addCompassTarget(holder.uniqueId, targetName)

            val currentLore = meta.lore
            val trackingLore = ChatColor.translateAlternateColorCodes('&', "&7Tracking: &a${targetName}")
            if (currentLore == null || !currentLore.contains(trackingLore)) {
                meta.setLore(listOf(
                    trackingLore,
                    ChatColor.translateAlternateColorCodes('&', "&8Tracker Compass")
                ))
                compass.itemMeta = meta
            }
        }
    }
}
